### PR TITLE
[Cult 4] Rituals System, Return of Sacrifices, Cursed Blood

### PR DESCRIPTION
**Pull request by gurfan**
**Creates a rituals system for cultists.  Heavy rough draft.**
The goal of this is to start implementing the planned changes that were listed in
https://docs.google.com/document/d/1e49TxEMFdzAwpvIy6alBgYrytWfAo5F-O5A62PW2KvY/edit
and 
https://docs.google.com/document/d/1rlOxtkmFLOlVJVRo1jeVumdv8Tc_VcdAIqkk-0BwvOQ/edit

The idea of the rituals system is to give cultists a series of nonlinear objectives to complete throughout the round.
By completing rituals, cultists can earn tattoos for themselves. Completing ritual also adds to a hidden "score" which they'll be able to see on the round-end screen (the "score" will probably have more significance later on but that is outside of this PR, ex. gating spells behind a certain "score")

Cultists start with three unlocked rituals, and get a new ritual for each one completed. Most rituals can be completed multiple times for extra points/tattoos.

The current rituals are pretty simple and not all tattoos are obtainable yet; please give suggestions for more rituals/rewards/tweaks/etc

**TWEAKS**
- Pylons can now be built be human cultists using the raise structure spell.
- Humans can now be sacrificed at an altar again to spawn a bloodstone. Sacrificing someone at an altar requires two cultists and can be done at any time. The bloodstone cultifies an area around it like before but it won't summon monsters when attacked.
- Blood Daggers can now be sheathed and unsheathed from a MindUI element on the HUD, instead of a spell. Xenomorphs can wield blood daggers (if they can somehow get the tattoo)

**Skin Rune (Runic Skin? Rune Tattoo?)**
A new tattoo granted to cultists who successfully convert a crewmember. Allows you to store a talisman in your skin and activate it whenever it is needed.

**Cursed Blood**
Blood that touches a tempting goblet (created by paraphernalia rune or soul-stoned people) becomes "cursed". The "curse" is a special type of virus. It can be cured through normal virology vaccines or by drinking holy water and waiting.  
Curse Symptoms: 
- Victim mutters cult phrases and sees runes drawn around them.
- Victim occasionally vomits up large amounts of blood.
- Victim occasionally gets sent through the veil as if travelling with a path rune. The place they get sent to is random (excluding dangerous places like solars, xenobio, the engine) 
- Victim is turned into a shadow person. After enough time they eventually turn into a faithless.
These effects won't activate if the victim is standing in the chapel, or if the victim is a cultist.

Current Rituals Implemented:
- Draw Full Rune (Always available and hidden, only contributes to roundend score, score lowered when rune is erased)
- Sow Confusion (Use confusion runes at least 2 people)
- Sacrifice Animal (Sacrifice small animals, any animal bigger than a mouse grants the blood dagger tattoo)
- Spirited Away (Send a non-cultist through a path rune, grants shortcut tracer tattoo)
- Sacrifice Human (Sacrifice a human at an altar, grants pale body tattoo)
- Reveal The Truth (Trap non-cultists in revealed runes)
- Silence The Lambs (Use Deaf/Mute on at least 2 people)
- Spread The Gift (Curse at least 3 people with cursed blood)
- Blood Missionary (Convert a crewmember, grants skin rune tattoo)
- Forge Construct (Use a soulstone on a construct shell)
- Spill Blood (Spread blood across 4% of station tiles)
More rituals to be added soon.

TODO:

- [x] Talking with Nar'sie gives a list of valid rituals
- [x] Offerings Altar, Veil Weakener, etc: structure that can be built by cultists, cultifies area around it and does other things. (need spritework) (Probably for a future PR)
- [x] More rituals
- [x] Let chaplains bless people to remove curse
- [x] Other stuff

**Old Videos:**

https://user-images.githubusercontent.com/51427771/160915951-38356250-85b3-4684-8aaf-32fe01fc423f.mp4

https://user-images.githubusercontent.com/51427771/160916242-d91f49eb-52cd-453c-a521-788da68d89ba.mp4

(Has since been changed to look smoother and not activate in the chapel)

:cl:
 * rscadd: Cult Rituals have been added to give cultists some objectives. Cultists are given three at round start, and another for each ritual completed. Completing rituals will contribute to a round-end score and sometimes grant a tattoo.
 * rscadd: Cultists can now sacrifice people again to create a bloodstone. They can do so at any time now, but sacrifices now require at least two cultists.
 * rscadd: Cultists can now create cursed blood by pouring it into a tempting goblet. The cursed blood will infect any non-cultist who ingests it, causing a variety of ill effects. The curse can be cured with holy water, blessing by a chaplain, or through virology.
 * tweak: Pylons can now be built by cultists using the raise structure rune.

### CODE REVIEW

**Comment by Servalfang** [comment]
its GULT

**Comment by Optimism333** [comment]
ayyyyyyyyyyy

**Comment by Eneocho** [comment]
Very nice. The spirit of Deity lives on.

**Comment by Eneocho** [comment]
You linked the same google docs file twice.
Did you mean to link: https://docs.google.com/document/d/1e49TxEMFdzAwpvIy6alBgYrytWfAo5F-O5A62PW2KvY ?

**Comment by BurntDevil** [comment]
if only all antags had a progression structure to follow like this. i believe thats one of the core issues with gameplay, and especially dynamic. so this is a good step forward

**Comment by Spaceman33333** [comment]
Without deity cult may even be good. It's time to bring it back.

**Comment by Eneocho** [comment]
Can you make it so solars don't make power during the eclipse? For muh immulsions.

**Comment by Eneocho** [comment]
You could change the eclipse message from a generic "the star disappeared, be careful for the next 10m" to a more terrified/desperate "shit's hitting the fan real fast, you guys better stop the end of the universe as we know it".

**Comment by gurfan** [comment]
Im not adding the eclipse in this PR

**Comment by Eneocho** [comment]
Oh, noted. But still, consider it for when you do, please.

**Comment by Cartlord** [comment]
do you plan to re-add some kind of god summoning, whether it's nar'sie or something gunkier like the honkmother?

**Comment by Eneocho** [comment]
Yes, assuming he follows the linked google drive files, the eclipse thing will be a period where the cult can either summon narnar herself or get a turbo augmented cultist to become a living avatar of narnar. I would love some clown-themed cult, or a clockwork themed one. Keep everything the same, but change the icons depending on the cult type the random selection chose.

**Comment by gurfan** [comment]
This is ready for review. The rituals are still really basic but I'm going to leave bigger additions to future PR's.

File: code/_onclick/mindUI/bloodcult/cultist.dm [hunk 1]
```diff
@@ -144,6 +145,239 @@
 
 	move_whole_ui = TRUE
 
+
+
+////////////////////////////////////////////////////////////////////
+//																  //
+//					BLOODCULT - LEFT PANEL						  //
```
Comment (DamianX) [inline code/_onclick/mindUI/bloodcult/cultist.dm]:
```suggestion
//					BLOODCULT - RIGHT PANEL						  //
```

File: code/datums/gamemode/factions/bloodcult/bloodcult.dm [hunk 2]
```diff
@@ -26,20 +26,47 @@
...
+			return TRUE
 
 /datum/faction/bloodcult/IsSuccessful()
 	return cult_win
 
 
+/datum/faction/bloodcult/GetScoreboard()
+	. = ..()
+	. += "<BR>Total Veil Weakness:[veil_weakness]<BR>"
+	return .
```
Comment (DamianX) [inline code/datums/gamemode/factions/bloodcult/bloodcult.dm]:
```suggestion

```

File: code/datums/gamemode/factions/bloodcult/bloodcult.dm [hunk 3]
```diff
@@ -78,12 +105,26 @@
...
+
+	if(!usr.check_rights(R_ADMIN))
+		message_admins("[usr] tried to access bloodcult faction Topic() without permissions.")
+		return
+
+	if(href_list["unlockRitual"])
+		var/datum/bloodcult_ritual/R = input(usr,"Select a ritual to unlock.", "Unlock", null) as null|anything in locked_rituals
+		if(R)
+			R.Unlock(TRUE)
+			locked_rituals -= R
```
Comment (DamianX) [inline code/datums/gamemode/factions/bloodcult/bloodcult.dm]: This should be done inside `Unlock()`, no?

File: code/datums/gamemode/factions/bloodcult/bloodcult.dm [hunk 4]
```diff
@@ -100,4 +142,74 @@
 		for (var/datum/role/R in members)
 			var/mob/M = R.antag.current
 			to_chat(M, "<span class='sinister'>This number might rise up to 9 as more people arrive aboard the station.</span>")
+	AnnounceObjectives()
 	..()
+
+
+/datum/faction/bloodcult/proc/GetDepartmentName(var/area/D)
```
Comment (DamianX) [inline code/datums/gamemode/factions/bloodcult/bloodcult.dm]: This proc seems unused and the implementation seems wrong. The switch only checks if the types are equal, subtypes are not considered equals. I think.
Remove this if it's unused, fix it and use it if it's meant to be used.

File: code/datums/gamemode/factions/bloodcult/bloodcult.dm [hunk 5]
```diff
@@ -100,4 +142,74 @@
...
+		if(/area/supply)
+			return "Cargo"
+		if(/area/medical)
+			return "Medical"
+		if(/area/engineering)
+			return "Engineering"
+		else
+			return D.name
+
+/datum/faction/bloodcult/proc/IsValidDepartment(var/area/D)
```
Comment (DamianX) [inline code/datums/gamemode/factions/bloodcult/bloodcult.dm]: This proc seems unused as well

File: code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm [hunk 6]
```diff
@@ -368,28 +392,30 @@
...
+			L.resting = 1
+			L.update_canmove()
 
-		add_fingerprint(C)
+		add_fingerprint(L)
 
 	O.forceMove(loc)
-	to_chat(user, "<span class='warning'>You move \the [O] on top of \the [src]</span>")
+	if(O == user)
+		to_chat(user, "<span class='warning'>You climb on top of \the [src]</span>")
```
Comment (DamianX) [inline code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm]:
```suggestion
		to_chat(user, "<span class='warning'>You climb on top of \the [src].</span>")
```

File: code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm [hunk 7]
```diff
@@ -368,28 +392,30 @@
...
+		user.resting = 1
+	else
+		to_chat(user, "<span class='warning'>You move \the [O] on top of \the [src]</span>")
```
Comment (DamianX) [inline code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm]:
```suggestion
		to_chat(user, "<span class='warning'>You move \the [O] on top of \the [src].</span>")
```

File: code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm [hunk 8]
```diff
@@ -368,28 +392,30 @@
...
+			L.update_canmove()
 
-		add_fingerprint(C)
+		add_fingerprint(L)
 
 	O.forceMove(loc)
-	to_chat(user, "<span class='warning'>You move \the [O] on top of \the [src]</span>")
+	if(O == user)
+		to_chat(user, "<span class='warning'>You climb on top of \the [src]</span>")
+		user.resting = 1
```
Comment (DamianX) [inline code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm]:
```suggestion
		user.resting = TRUE
```

File: code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm [hunk 9]
```diff
@@ -368,28 +392,30 @@
...
+		if (!do_after(user,L,15))
 			return
-		C.unlock_from()
+		L.unlock_from()
 
-		if (ishuman(C))
-			C.resting = 1
-			C.update_canmove()
+		if (ishuman(L) && L != user)
+			L.resting = 1
```
Comment (DamianX) [inline code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm]:
```suggestion
			L.resting = TRUE
```

File: code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm [hunk 10]
```diff
@@ -459,91 +485,61 @@
...
-		to_chat(user, "You remove \the [blade] from \the [src]</span>")
-		blade = null
-		playsound(loc, 'sound/weapons/blade1.ogg', 50, 1)
-		update_icon()
+		if(is_locking(lock_type))
+			if (do_after(user,src,20))
+				unlock_atom(get_locked(lock_type)[1])
+				blade.forceMove(loc)
+				blade.attack_hand(user)
+				to_chat(user, "You remove \the [blade] from \the [src]</span>")
```
Comment (DamianX) [inline code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm]:
```suggestion
				to_chat(user, "<span class='notice'>You remove \the [blade] from \the [src]</span>")
```

File: code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm [hunk 11]
```diff
@@ -459,91 +485,61 @@
...
+				blade = null
+				playsound(loc, 'sound/weapons/blade1.ogg', 50, 1)
+				update_icon()
+		else 
+			blade.forceMove(loc)
+			blade.attack_hand(user)
+			to_chat(user, "You remove \the [blade] from \the [src]</span>")
```
Comment (DamianX) [inline code/datums/gamemode/factions/bloodcult/bloodcult_buildings.dm]:
```suggestion
			to_chat(user, "<span class='notice'>You remove \the [blade] from \the [src]</span>")
```

File: code/game/gamemodes/scoreboard.dm [hunk 12]
```diff
@@ -0,0 +1,725 @@
...
+					continue
+				robolist += "[robo.name][robo.stat?" (Deactivated) (Played by: [get_key(robo)]), ":" (Played by: [get_key(robo)]), "]"
+			ai_completions += "[robolist]"
+
+	for (var/mob/living/silicon/robot/robo in mob_list)
+		if(!robo)
+			continue
+		var/icon/flat = getFlatIcon(robo)
+		if (!robo.connected_ai)
+			if (robo.stat != 2)
```
Comment (DamianX) [inline code/game/gamemodes/scoreboard.dm]:
```suggestion
			if (robo.stat != DEAD)
```

File: code/game/turfs/turf_flick_animations.dm [hunk 13]
```diff
@@ -47,7 +47,7 @@
 	if(invis)
 		animation.invisibility = invis
 	animation.icon = a_icon
-	animation.animate_movement = 0
+	animation.animate_movement = TRUE
```
Comment (DamianX) [inline code/game/turfs/turf_flick_animations.dm]: Did you mean 
```suggestion
	animation.animate_movement = animate_movement
```
?

File: code/datums/gamemode/factions/bloodcult/bloodcult.dm (hunk 3 above)
Comment (gurfan) [reply code/datums/gamemode/factions/bloodcult/bloodcult.dm]: It wasn't done inside Unlock(). Now it is.

**Pull request merged**